coordinator/src/admin.rs

```diff
@@ -144,6 +156,98 @@ pub async fn list_dlc_channels(
     Ok(Json(dlc_channels))
 }
 
+pub async fn collaborative_revert(
+    State(state): State<Arc<AppState>>,
+    revert_params: Json<CollaborativeRevert>,
+) -> Result<Json<String>, AppError> {
+    let channel_id_string = revert_params.channel_id.clone();
+    let channel_id = hex::decode(channel_id_string.clone())
+        .map_err(|err| AppError::BadRequest(err.to_string()))?;
+    let channel_id: [u8; 32] = channel_id
+        .try_into()
+        .map_err(|_| AppError::BadRequest("Provided channel ID was invalid".to_string()))?;
+
+    let mut conn =
+        state.pool.clone().get().map_err(|e| {
+            AppError::InternalServerError(format!("Failed to acquire db lock: {e:#}"))
+        })?;
+
+    let channel_details = state
+        .node
+        .inner
+        .channel_manager
+        .get_channel_details(&channel_id)
+        .context("Could not get channel")
+        .map_err(|error| {
+            AppError::InternalServerError(format!("No ln channel found: {error:#}"))
+        })?;
+
+    let sub_channels = state.node.inner.list_dlc_channels().map_err(|e| {
+        AppError::InternalServerError(format!("Failed to list DLC channels: {e:#}"))
+    })?;
+    let subchannel = sub_channels
+        .iter()
+        .find(|c| c.channel_id == channel_id)
+        .context("Could not find provided channel")
+        .map_err(|e| AppError::BadRequest(format!("Channel not found: {e:#}")))?;
+
+    let position = Position::get_position_by_channel_id(&mut conn, channel_id_string.clone())
+        .map_err(|error| {
+            tracing::error!(
+                channel_id = revert_params.channel_id.clone(),
+                "Could not get position for channel {error:#}"
+            );
+            AppError::InternalServerError(format!("Failed to load position from db: {error:#}"))
+        })?;
+
+    let settlement_amount = position
+        .calculate_settlement_amount(revert_params.price)
+        .map_err(|error| {
+            AppError::InternalServerError(format!("Could not calculate pnl {error:#}"))
+        })?;
+
+    // the coordinator gets the channel value - trader's balance - settlement_amount, i.e. the
+    // trader loses the tx fee locked up in the sub_channel transactions
+    // TODO: share the transaction fees between both parties
+    let coordinator_amount = subchannel.fund_value_satoshis as i64
+        - (channel_details.inbound_capacity_msat / 1000) as i64
+        - settlement_amount as i64;
+    let trader_amount = subchannel.fund_value_satoshis - coordinator_amount as u64;
+
+    let fee = weight_to_fee(
+        COLLABORATIVE_REVERT_TX_WEIGHT,
+        revert_params.fee_rate_sats_vb,
+    )
+    .expect("To be able to calculate constant fee rate");
+    let address = state.node.inner.get_unused_address();
+    let coordinator_amount = Amount::from_sat(coordinator_amount as u64 - fee / 2);
```

Comment:
🔧 what if the fee is uneven? one side should floor and the other ceil.

Reply:
We will pay 1 sat more to the miners which I think is ok